### PR TITLE
feat(nuxt): warn when using data-fetching composables outside setup/plugin

### PR DESCRIPTION
**Pull request by atinux**
### ❓ Type of change

- [x] 📖 Documentation (updates to the documentation, readme or JSdoc annotations)
- [x] 👌 Enhancement (improving an existing functionality like performance)

### 📚 Description

Added JSDoc comments for a better DX:

| before | after |
| - | - |
| CleanShot 2024-01-05 at 16 06 22@2x | CleanShot 2024-01-05 at 16 01 26@2x |
| CleanShot 2024-01-05 at 16 06 26@2x | CleanShot 2024-01-05 at 16 01 32@2x |

Added a warning when `useFetch` or `useAsyncData` is used on client-side on a component already mounted (or instance unavailable):

CleanShot 2024-01-05 at 16 05 20@2x

**Notes:**

It looks good on my tests in the playground, but before merging it want to test it on bigger projects to make sure I did not forgot anything.

### 📝 Checklist

- [ ] I have added tests (if possible).

### CODE REVIEW

File: packages/nuxt/src/app/composables/fetch.ts [hunk 1]
```diff
@@ -131,7 +143,8 @@ export function useFetch<
     immediate,
     getCachedData,
     deep,
-    watch: watch === false ? [] : [_fetchOptions, _request, ...(watch || [])]
+    watch: watch === false ? [] : [_fetchOptions, _request, ...(watch || [])],
+    _useFetch: true
```
Comment (pi0) [inline packages/nuxt/src/app/composables/fetch.ts:147]: One solution to hide it is to use a Symbol

Comment (Atinux) [reply packages/nuxt/src/app/composables/fetch.ts:147]: Tell me how 😄

File: packages/nuxt/src/app/composables/asyncData.ts [hunk 2]
```diff
@@ -259,6 +326,9 @@ export function useAsyncData<
   if (import.meta.client) {
     // Setup hook callbacks once per instance
     const instance = getCurrentInstance()
+    if (!instance || instance?.isMounted) {
```
Comment (danielroe) [inline packages/nuxt/src/app/composables/asyncData.ts:329]:
```suggestion
    if (import.meta.dev && (!instance || instance?.isMounted)) {
```

File: packages/nuxt/src/app/composables/fetch.ts (hunk 1 above)
Comment (danielroe) [reply packages/nuxt/src/app/composables/fetch.ts:147]: we could also avoid typing it entirely

**Comment by DamianGlowala** [comment]
Spotted in a few places:
'a SSR-friendly' -> 'a**n** SSR-friendly'

File: packages/nuxt/src/app/composables/asyncData.ts [hunk 3]
```diff
@@ -42,15 +42,51 @@ export interface AsyncDataOptions<
   PickKeys extends KeysOf<DataT> = KeysOf<DataT>,
   DefaultT = null,
 > {
+  /**
+   * Whether to fetch on server side.
```
Comment (DamianGlowala) [inline packages/nuxt/src/app/composables/asyncData.ts:46]:
```suggestion
   * Whether to fetch on the server side.
```

File: packages/nuxt/src/app/composables/asyncData.ts [hunk 4]
```diff
@@ -42,15 +42,51 @@ export interface AsyncDataOptions<
...
+   * When set to false, will prevent the request from firing immediately
+   * @default true
+   */
   immediate?: boolean
+  /**
+   * Return data in a deep ref object (it is true by default). It can be set to false to return data in a shallow ref object, which can improve performance if your data does not need to be deeply reactive.
+   */
   deep?: boolean
+  /**
+   * Avoid fetching same key more than once at a time
```
Comment (DamianGlowala) [inline packages/nuxt/src/app/composables/asyncData.ts:87]:
```suggestion
   * Avoid fetching the same key more than once at a time
```

File: packages/nuxt/src/app/composables/asyncData.ts [hunk 5]
```diff
@@ -113,6 +168,13 @@ export function useAsyncData<
   handler: (ctx?: NuxtApp) => Promise<ResT>,
   options?: AsyncDataOptions<ResT, DataT, PickKeys, DefaultT>
 ): AsyncData<PickFrom<DataT, PickKeys> | DefaultT, (NuxtErrorDataT extends Error | NuxtError ? NuxtErrorDataT : NuxtError<NuxtErrorDataT>) | null>
+/**
+ * Provides access to data that resolves asynchronously in a SSR-friendly composable.
+ * See {@link https://nuxt.com/docs/api/composables/use-async-data}
+ * @param key a unique key to ensure that data fetching can be properly de-duplicated across requests.
```
Comment (DamianGlowala) [inline packages/nuxt/src/app/composables/asyncData.ts:174]:
```suggestion
 * @param key A unique key to ensure that data fetching can be properly de-duplicated across requests.
```

File: packages/nuxt/src/app/composables/asyncData.ts [hunk 6]
```diff
@@ -102,6 +150,13 @@ export function useAsyncData<
...
```
Comment (DamianGlowala) [inline packages/nuxt/src/app/composables/asyncData.ts:156]:
```suggestion
 * @param key A unique key to ensure that data fetching can be properly de-duplicated across requests.
```

**Review by DamianGlowala — CHANGES_REQUESTED** [review]

**Review by DamianGlowala — CHANGES_REQUESTED** [review]

**Review by danielroe — APPROVED** [review]

**Comment by danielroe** [comment]
It shouldn't. Please raise an issue if you encounter an issue and I'll look into it straight away.

**Comment by HADB** [comment]
I'm using useFetch in a custom composable and will call this composable on each page mounted. I will get this warning when switching pages. Is this as expected?

**Comment by danielroe** [comment]
Indeed, it's warning you not to call the composable in the mounted hook.

**Comment by HADB** [comment]
I'm doing some common response code handling in my custom composable and which will be called both in or out the mounted hook. What should I use to call the apis if `useFetch` is not allowed to use in the custom composable?

**Comment by danielroe** [comment]
useFetch returns a refresh function that can be called at any time. If you don't want to run it in setup, you can pass `immediate: false` to defer running the function until a time of your choosing.

**Comment by dalirnet** [comment]
> useFetch returns a refresh function that can be called at any time. If you don't want to run it in setup, you can pass `immediate: false` to defer running the function until a time of your choosing.

This warning is also displayed when the refresh function is used.

**Comment by danielroe** [comment]
Are you sure about that? If so, would you raise an issue with a reproduction? 🙏

**Comment by matt-clegg** [comment]
Is there anyway of knowing where in the codebase this warning is being thrown from? We have updated Nuxt and spotted several of these warnings but it's not immediately clear which files/components are throwing the warning logs.

**Comment by AnthonyRuelle** [comment]
I have warnings everywhere with this version upgrade. I don't use useFetch in .vue files for the simple reason that I have a very large application and I split that via the repository pattern. What do you recommend for sharing API call code without duplicating everything .. ?

**Comment by manniL** [comment]
@matt-clegg set a debugger at the warning or use console.trace there (locally) ☺️ 

@AnthonyRuelle This is better suited in a discussion including a reproduction :)

**Comment by AnthonyRuelle** [comment]
> @matt-clegg set a debugger at the warning or use console.trace there (locally) ☺️ 
> 
> @AnthonyRuelle This is better suited in a discussion including a reproduction :)


Yes of course, I created a discussion with an example in my project

https://github.com/nuxt/nuxt/discussions/25625

**Pull request merged**